### PR TITLE
Invoke-DbaQuery - Accept Microsoft.Data.SqlClient.SqlParameter for SqlParameter

### LINKED ISSUE #7434: Support SqlParameter class for Invoke-DbaAsync SqlParameters argument
### Summary of new feature

* Support initiliaze class sqlparameter as stored procedure parameter. It's can be useful to execute a stored procedure which returns result via output variable. 

### Proposed technical details (if applicable)

* https://docs.microsoft.com/ru-ru/dotnet/api/system.data.sqlclient.sqlparameter?view=dotnet-plat-ext-5.0
* https://docs.microsoft.com/ru-ru/dotnet/api/system.data.sqlclient.sqlparametercollection.add?view=dotnet-plat-ext-5.0#System_Data_SqlClient_SqlParameterCollection_Add_System_Data_SqlClient_SqlParameter_

Something like:
```
CREATE OR ALTER PROC [dbo].[my_proc]
    @json_result nvarchar(max) output
AS
BEGIN
  set @json_result = (
      select 'sample' as 'example'
      for json path, without_array_wrapper
  );
END
```

```
$QueryParameters = @{}

$outputParameter = New-Object System.Data.SqlClient.SqlParameter -Property @{
    ParameterName = "@json_result";
    SqlDbType = [System.Data.SqlDbType]::NVarChar;
    Size = -1;
    Direction = [System.Data.ParameterDirection]::Output;
};

$QueryParameters.Add("json_result", $outputParameter) # some unnecessary line

Invoke-DbaQuery -SqlInstance $sqlInstance -Database test -CommandType StoredProcedure -Query "my_proc" -SqlParameters $QueryParameters > $null

$outputParameter
```

### Latest version of dbatools as of writing

* 1.0.156

### PR DESCRIPTION
**Pull request by potatoqualitee**
## Type of Change
 - [x] New feature (non-breaking change, adds functionality, fixes #7434 )

* Now accepts both hashtables and Microsoft.Data.SqlClient.SqlParameter
* Created new command to make creating SqlParameters easier
* Changed and aliased SqlParameter since it was plural and PowerShell best practices says singular

@i7slegend if you can confirm that this is what you were asking for, I'd appreciate. I don't use Invoke-DbaQuery much and minimally tested

```powershell
$output = New-DbaSqlParameter -ParameterName json_result -SqlDbType NVarChar -Size -1 -Direction Output
Invoke-DbaQuery -SqlInstance localhost -Database master -CommandType StoredProcedure -Query my_proc -SqlParameters $output
$output.Value
```

```powershell
$output = New-DbaSqlParameter -ParameterName json_result -SqlDbType NVarChar -Size -1 -Direction Output
$hash = @{ whatever = $output }
Invoke-DbaQuery -SqlInstance localhost -Database master -CommandType StoredProcedure -Query my_proc -SqlParameters $hash
$output.Value
```

### CODE REVIEW

File: functions/Invoke-DbaQuery.ps1 [hunk 1]
```diff
@@ -115,9 +115,16 @@ function Invoke-DbaQuery {
...
             "EndDate" = $enddate;
         };
-        PS C:\> Invoke-DbaQuery -SqlInstance "server1" -Database tempdb -Query "Example_SP" -SqlParameters $QueryParameters -CommandType StoredProcedure
+        PS C:\> Invoke-DbaQuery -SqlInstance "server1" -Database tempdb -Query "Example_SP" -SqlParameter $QueryParameters -CommandType StoredProcedure
 
         Executes a stored procedure Example_SP using multiple SQL Parameters
+
+    .EXAMPLE
+        PS C:\> $output = New-DbaSqlParameter -ParameterName json_result -SqlDbType NVarChar -Size -1 -Direction Output
+        PS C:\> Invoke-DbaQuery -SqlInstance localhost -Database master -CommandType StoredProcedure -Query my_proc -SqlParameters $output
```
Comment (andreasjordan) [inline functions/Invoke-DbaQuery.ps1:124]: -SqlParameter

Comment (potatoqualitee) [reply functions/Invoke-DbaQuery.ps1:124]: Thanks! btw, refresh, i just updated a bunch of stuff

**Comment by niphlod** [comment]
not entirely happy with the breaking change of sqlparameter --> sqlparameters. This stems from the original Invoke-SqlCmd2 syntax (which sooner or later we'd like to contribute back to)

**Comment by niphlod** [comment]
whoops, I see it's aliased. okay. 
Problem here is that people may wanna pass an arbitrary number of params, whichever type

File: functions/New-DbaSqlParameter.ps1 [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        PS C:\> $output.Value
+
+        Creates an output parameter and uses it to invoke a stored procedure.
+    #>
+    [CmdletBinding()]
+    param(
+        [ValidateSet("None", "IgnoreCase", "IgnoreNonSpace", "IgnoreKanaType", "IgnoreWidth", "BinarySort2", "BinarySort")]
+        [string]$CompareInfo,
+        [ValidateSet("AnsiString", "Binary", "Byte", "Boolean", "Currency", "Date", "DateTime", "Decimal", "Double", "Guid", "Int16", "Int32", "Int64", "Object,
+        SByte", "Single", "String", "Time", "UInt16", "UInt32", "UInt64", "VarNumeric", "AnsiStringFixedLength", "StringFixedLength", "Xml,
```
Comment (niphlod) [inline functions/New-DbaSqlParameter.ps1:104]: not sure this is indented correctly

File: functions/New-DbaSqlParameter.ps1 [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        [string]$ParameterName,
+        [string]$Precision,
+        [string]$Scale,
+        [int]$Size,
+        [string]$SourceColumn,
+        [switch]$SourceColumnNullMapping,
+        [ValidateSet("Original", "Current", "Proposed", "Default")]
+        [string]$SourceVersion,
+        [ValidateSet("BigInt", "Binary", "Bit", "Char", "DateTime", "Decimal", "Float", "Image", "Int", "Money", "NChar", "NText", "NVarChar", "Real,
+        UniqueIdentifier", "SmallDateTime", "SmallInt", "SmallMoney", "Text", "Timestamp", "TinyInt", "VarBinary", "VarChar", "Variant", "Xml,
```
Comment (niphlod) [inline functions/New-DbaSqlParameter.ps1:123]: not sure this is indented correctly

File: functions/New-DbaSqlParameter.ps1 [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        [Alias("Name")]
+        [string]$ParameterName,
+        [string]$Precision,
+        [string]$Scale,
+        [int]$Size,
+        [string]$SourceColumn,
+        [switch]$SourceColumnNullMapping,
+        [ValidateSet("Original", "Current", "Proposed", "Default")]
+        [string]$SourceVersion,
+        [ValidateSet("BigInt", "Binary", "Bit", "Char", "DateTime", "Decimal", "Float", "Image", "Int", "Money", "NChar", "NText", "NVarChar", "Real,
```
Comment (niphlod) [inline functions/New-DbaSqlParameter.ps1:122]: not sure this is indented correctly

**Review by niphlod — CHANGES_REQUESTED** [review]

**Comment by niphlod** [comment]
all things considered, I'd vouch for Sqlparameters to stay as dict, mappings are always by name. Then check each value of the dict for being "simple" or a full fledged New-DbaSqlParameter

File: functions/New-DbaSqlParameter.ps1 (hunk 3 above)
Comment (potatoqualitee) [reply functions/New-DbaSqlParameter.ps1:123]: you were right! thanks, that was super messed up.

File: functions/New-DbaSqlParameter.ps1 (hunk 2 above)
Comment (potatoqualitee) [reply functions/New-DbaSqlParameter.ps1:104]: you were right! thanks, that was super messed up.

**Comment by potatoqualitee** [comment]
@niphlod im hearing you are concerned about validation. I will add in some extra validation 👍🏼

**Comment by potatoqualitee** [comment]
Added validation 

image

image

**Comment by niphlod** [comment]
validation apart, the interface as it is is clunky: accepting a list of full fledged params OR a dict IMHO can be better designed.
What I'd strive for is a dict whose VALUES can be either simple types or full-fledged params.

**Comment by potatoqualitee** [comment]
All tests pass and I think I addressed what legend was looking for 💯

**Pull request merged**